File: frontend/src/types/sentiment.ts
import type { SignalType } from "./stock";

export interface AlphaMetric {
  id: string;
  stock_id: string | null;
  ticker: string | null;
  company_name: string | null;
  sector: string | null;
  expectation_gap: number;
  narrative_velocity: number;
  sentiment_divergence: number;
  composite_score: number;
  signal: SignalType;
  conviction: number;
  computed_at: string;
}

export interface SentimentOverview {
  market_sentiment: number;
  bullish_count: number;
  bearish_count: number;
  neutral_count: number;
  total_articles: number;
  sentiment_trend: SentimentTrendPoint[];
  sectors: SectorSentiment[];
  updated_at: string;
}

export interface SentimentTrendPoint {
  timestamp: string;
  sentiment: number;
  volume: number;
}

export interface SectorSentiment {
  sector: string;
  sentiment_score: number;
  news_count: number;
  top_signal: SignalType;
  change_24h: number;
}

export interface SentimentAnalysis {
  article_id: string;
  sentiment_score: number;
  confidence: number;
  explanation: string | null;
  impact_timeline: string;
  entities: string[];
  topics: string[];
}

export interface CompositeSignal {
  ticker: string;
  company_name: string;
  signal: SignalType;
  composite_score: number;
  conviction: number;
  components: {
    expectation_gap: number;
    narrative_velocity: number;
    sentiment_divergence: number;
  };
  reasoning: string | null;
  computed_at: string;
}

File: backend/app/schemas/sentiment.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel


class SentimentResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    article_id: uuid.UUID
    sentiment_score: float
    confidence: float
    explanation: str | None = None
    impact_timeline: str
    finbert_score: float | None = None
    llm_score: float | None = None
    llm_provider: str | None = None
    analyzed_at: datetime


class AlphaMetricResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    stock_id: uuid.UUID | None = None
    sector: str | None = None
    expectation_gap: float
    narrative_velocity: float
    sentiment_divergence: float
    composite_score: float
    signal: str
    conviction: float
    computed_at: datetime
    window_hours: int


class SentimentOverview(BaseModel):
    market_sentiment: float
    bullish_count: int
    bearish_count: int
    neutral_count: int
    top_movers: list[AlphaMetricResponse]


class SectorSentiment(BaseModel):
    sector: str
    sentiment_score: float
    article_count: int
    top_stocks: list[str]
